File: api/src/schemas/admin/seo.py
"""SEO 설정 스키마 — 관리자 설정 → SEO.

브라우저 탭 제목·검색엔진 노출 텍스트·카카오톡/SNS 미리보기 이미지·즐겨찾기 아이콘을
관리자 화면에서 직접 편집한다.

필드:
- site_name        브라우저 탭 제목 + Open Graph 기본 제목
- site_description 검색 결과/공유 미리보기 설명
- keywords         검색엔진 키워드 (쉼표 구분, Pydantic은 문자열로 받고 서비스에서 정리)
- og_image_url     카카오톡/페이스북 공유 시 보이는 1200×630 권장 이미지 URL
- favicon_url      브라우저 탭 옆 작은 아이콘 URL

응답은 default_* 도 같이 돌려준다 → 관리자 화면에서 "기본값으로 초기화" 안내용.
"""

from pydantic import BaseModel, Field


class SeoConfigResponse(BaseModel):
    site_name: str
    site_description: str
    keywords: str
    og_image_url: str
    favicon_url: str
    default_site_name: str
    default_site_description: str
    default_keywords: str
    default_og_image_url: str
    default_favicon_url: str


class SeoConfigUpdateRequest(BaseModel):
    site_name: str = Field(min_length=1, max_length=120)
    site_description: str = Field(min_length=1, max_length=300)
    # keywords 는 빈 문자열 허용 (검색엔진 키워드는 선택사항).
    keywords: str = Field(max_length=300, default="")
    og_image_url: str = Field(min_length=1, max_length=500)
    favicon_url: str = Field(min_length=1, max_length=500)


class SeoAssetUploadResponse(BaseModel):
    """업로드 후 응답 — URL 을 SeoConfigUpdateRequest 의 해당 필드에 넣어 PUT."""

    asset_url: str
    kind: str  # "favicon" | "og_image"

File: web/src/features/admin-dashboard/api/seo.ts
const API_BASE = process.env.NEXT_PUBLIC_API_URL ?? "http://localhost:8000";

export interface SeoConfig {
  site_name: string;
  site_description: string;
  keywords: string;
  og_image_url: string;
  favicon_url: string;
  default_site_name: string;
  default_site_description: string;
  default_keywords: string;
  default_og_image_url: string;
  default_favicon_url: string;
}

export interface SeoConfigUpdateInput {
  site_name: string;
  site_description: string;
  keywords: string;
  og_image_url: string;
  favicon_url: string;
}

export interface SeoAssetUploadResult {
  asset_url: string;
  kind: "favicon" | "og_image";
}

async function parseError(res: Response, fallback: string): Promise<Error> {
  let detail = `${fallback} (HTTP ${res.status})`;
  try {
    const body = (await res.json()) as {
      detail?: string | { message?: string } | unknown;
      message?: string;
    };
    if (typeof body?.detail === "string") {
      detail = body.detail;
    } else if (
      body?.detail &&
      typeof body.detail === "object" &&
      "message" in (body.detail as Record<string, unknown>) &&
      typeof (body.detail as { message?: unknown }).message === "string"
    ) {
      detail = (body.detail as { message: string }).message;
    } else if (typeof body?.message === "string") {
      detail = body.message;
    }
  } catch {
    // ignore
  }
  return new Error(detail);
}

export async function fetchSeoConfig(): Promise<SeoConfig> {
  const res = await fetch(`${API_BASE}/api/v1/admin/seo-config`, {
    credentials: "include",
  });
  if (!res.ok) throw await parseError(res, "SEO 설정을 불러오지 못했습니다");
  return res.json() as Promise<SeoConfig>;
}

export async function updateSeoConfig(
  input: SeoConfigUpdateInput,
): Promise<SeoConfig> {
  const res = await fetch(`${API_BASE}/api/v1/admin/seo-config`, {
    method: "PUT",
    credentials: "include",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(input),
  });
  if (!res.ok) throw await parseError(res, "SEO 설정 저장에 실패했습니다");
  return res.json() as Promise<SeoConfig>;
}

export async function uploadSeoAsset(
  kind: "favicon" | "og_image",
  file: File,
): Promise<SeoAssetUploadResult> {
  const formData = new FormData();
  formData.append("file", file);

  const res = await fetch(
    `${API_BASE}/api/v1/admin/seo-config/asset-upload?kind=${kind}`,
    {
      method: "POST",
      credentials: "include",
      body: formData,
    },
  );
  if (!res.ok) throw await parseError(res, "이미지 업로드에 실패했습니다");
  return res.json() as Promise<SeoAssetUploadResult>;
}
